### ISSUE TITLE
Failed to emerge sys-libs/glibc-2.31-r6

### ISSUE THREAD

**Issue by addeps3** [InBetweenNames/gentooLTO#632]
Hi, this a new gentoo system, installed via another installed distro (on to a different hdd)

There is no DE installed, and after following gentooLTO instructions, it was time to rebuild with 
lto-rebuild -r
There where circa eight or nine packages to be rebuilt and build failed on last one (glibc)

Here is output of `emerge --info '=sys-libs/glibc-2.31-r6::gentoo`

https://www.dropbox.com/s/brcemw33ezt5vbn/inf?dl=0

Output of 
`emerge -pqv =sys-libs/glibc-2.31-r6::gentoo`

`[ebuild   R   ] sys-libs/glibc-2.31-r6  USE="(crypt) multiarch (multilib) ssp (static-libs) -audit -caps (-cet) -compile-locales -custom-cflags -doc -gd -headers-only -nscd -profile (-selinux) -static-pie -suid -systemtap -test (-vanilla)" 
`

And here is the build log

https://www.dropbox.com/s/y9iuutrvzhxk9ex/build.log?dl=0

ebuild environment: 

https://www.dropbox.com/s/uflwfhq6zq7pklv/eb?dl=0

**Comment by telans** [comment]
Your glibc isn't being built with custom flags anyway (`-custom-cflags` USE).
```
[32;01m*[0m CFLAGS='-march=native -O3 -pipe -Wl,-O1 -Wl,--as-needed'
[32;01m*[0m CXXFLAGS='-march=native -O3 -pipe -Wl,-O1 -Wl,--as-needed'
```

Aside from that this repo is intended to be used with ~amd64, for which glibc-2.3-r2 builds fine

**Comment by addeps3** [comment]
> Your glibc isn't being built with custom flags anyway (`-custom-cflags` USE).
> 
> ```
> �[32;01m*�[0m CFLAGS='-march=native -O3 -pipe -Wl,-O1 -Wl,--as-needed'
> �[32;01m*�[0m CXXFLAGS='-march=native -O3 -pipe -Wl,-O1 -Wl,--as-needed'
> ```
> 
> According to your `emerge --info` you also don't have flags set in `make.conf`.
> 
> Aside from that this repo is intended to be used with ~amd64, for which glibc-2.32-r2 builds fine. This just seems like an issue with base gentoo

I havent set any package specific flags in make.conf because I didnt know gentoo would break if I didnt. 

Seems like an issue with base gentoo, ok, any suggestions what i could try right now?

**Comment by telans** [comment]
> Oh, do you mean that the make.conf.lto flags arent being used?

I do, if you haven't sourced `make.conf.lto` (or `make.conf.lto.defines`).

Otherwise you aren't using flags supported with this repo and I think lto-rebuild is just rebuilding your packages with original cflags. Not sure on what lto-rebuild is doing.

> Seems like an issue with base gentoo, ok, any suggestions what i could try right now?

Use '''unstable''' gentoo (~amd64, which is almost never unstable), `emerge -e @world`, follow the steps in the reade here to enable LTO/graphite etc, read the configuration files that `ltoize` installs and then `emerge -e @world --keep-going` to rebuild everything with LTO.

Seeing as this is a new installation, it may just be easier to start from scratch with ~amd64, then enable `ltoize` etc

**Comment by addeps3** [comment]
> > Oh, do you mean that the make.conf.lto flags arent being used?
> 
> I do, if you haven't sourced `make.conf.lto` (or `make.conf.lto.defines`) within `/etc/portage/make.conf`. This is all in the readme
> 
> Otherwise you aren't using flags supported with this repo and I think lto-rebuild is just rebuilding your packages with original cflags. Not sure on what lto-rebuild is doing.
> 
> > Seems like an issue with base gentoo, ok, any suggestions what i could try right now?
> 
> Use '''unstable''' gentoo (~amd64, which is almost never unstable), `emerge -e @world`, follow the steps in the readme here to enable LTO/graphite etc, read the configuration files that `ltoize` installs and then `emerge -e @world --keep-going` to rebuild everything with LTO.
> 
> Seeing as this is a new installation, it may just be easier to start from scratch with ~amd64, then enable `ltoize` etc

I should have just posted my make.conf because I did follow the readme, not saying I couldnt have made a mistake, but I *have* read it

`NTHREADS="6"

source make.conf.lto


COMMON_FLAGS="-march=native -O3 -pipe"
CFLAGS="${COMMON_FLAGS}"
CXXFLAGS="${COMMON_FLAGS}"
FCFLAGS="${COMMON_FLAGS}"
FFLAGS="${COMMON_FLAGS}"
MAKEOPTS="-j6"

# NOTE: This stage was built with the bindist Use flag enabled
PORTDIR="/var/db/repos/gentoo"
DISTDIR="/var/cache/distfiles"
PKGDIR="/var/cache/binpkgs"


LC_MESSAGES=C`

This _fucking_ website wont let me paste text as code

**Comment by telans** [comment]
> What do you mean ~amd64? This is a 64-bit system.

The tilde `~` represents using "testing" packages (esentially packages that are up to date, I thought it was called unstable). For example with [sys-libs/glibc](https://packages.gentoo.org/packages/sys-libs/glibc) you see those packages are marked with yellow. This is covered by the [ACCEPT_KEYWORDS](https://wiki.gentoo.org/wiki/ACCEPT_KEYWORDS) variable. I *really* recommend using this if you want any semblance of up to date packages.

Apologies, I assumed you hadn't sourced anything and just installed ltoize. The `make.conf.lto.defines` file specifies variables that contain certain flags (which itself is sourced within `make.conf.lto`). To enable these, you need to put the flags you wish to use within `COMMON_FLAGS`. Although it seems `make.conf.lto` already specifies CFLAGS at the bottom of the file, I believe since you are re-setting the variable within `make.conf` it's not picking up those flags. This is how I have it set up in my `make.conf` after sourcing `make.conf.lto.defines`:

```
NTHREADS="17"
source /etc/portage/make.conf.lto.defines
...
COMMON_FLAGS="-march=znver2 -O3 ${FLTO} -flto-compression-level=9 ${GRAPHITE} ${IPAPTA} ${SEMINTERPOS} ${DEVIRTLTO} -pipe" #${SAFER_FAST_MATH} -pipe
CFLAGS="${COMMON_FLAGS}"
CXXFLAGS="${COMMON_FLAGS}"
FFLAGS="${COMMON_FLAGS}"
FCFLAGS="${COMMON_FLAGS}"
OBJCFLAGS="${COMMON_FLAGS}"
OBJCXXFLAGS="${COMMON_FLAGS}"
RUSTFLAGS="-Ctarget-cpu=znver2 -Copt-level=3 -C codegen-units=17"
```

**Comment by addeps3** [comment]
> > What do you mean ~amd64? This is a 64-bit system.
> 
> The tilde `~` represents using "testing" packages (esentially packages that are up to date, I thought it was called unstable). For example with [sys-libs/glibc](https://packages.gentoo.org/packages/sys-libs/glibc) you see those packages are marked with yellow. This is covered by the [ACCEPT_KEYWORDS](https://wiki.gentoo.org/wiki/ACCEPT_KEYWORDS) variable. I _really_ recommend using this if you want any semblance of up to date packages - and it's what gentooLTO supports.
> 
> Apologies, I assumed you hadn't sourced anything and just installed ltoize. 

No problem at all! 

This is how Ive set up make.conf now, gonna run lto-rebuild and see how it goes. 

>NTHREADS="6"

>source make.conf.lto


>COMMON_FLAGS="-march=native -O3 -pipe ${GRAPHITE} ${DEVIRTLTO} ${IPAPTA} ${SEMINTERPOS} ${FLTO} -fuse-linker-plugin"
>CFLAGS="${COMMON_FLAGS}"
>CXXFLAGS="${COMMON_FLAGS}"
>FCFLAGS="${COMMON_FLAGS}"
>FFLAGS="${COMMON_FLAGS}"
>MAKEOPTS="-j6"

In case I go to sleep, have a good night. :)

**Comment by telans** [comment]
> COMMON_FLAGS="-march=native -O3 -pipe ${GRAPHITE}

I don't think you've mentioned if GCC has been built with the `graphite` USE flag or not, but it needs to be for the ${GRAPHITE} flags. And I'm not sure if you need to source `make.conf.lto.defines` directly to use those variable or not. You can veryify that they're being used at the beginning of each package emerge, where it prints variables in use such as the flags.

**Comment by addeps3** [comment]
> > COMMON_FLAGS="-march=native -O3 -pipe ${GRAPHITE}
> 
> I don't think you've mentioned if GCC has been built with the `graphite` USE flag or not, but it needs to be for the ${GRAPHITE} flags. And I'm not sure if you need to source `make.conf.lto.defines` directly to use those variables or not. You can verify that they're being used at the beginning of each package emerge, where it prints variables in use such as the flags.

make.conf.lto sources .defines itself, I saw no reason to change anything. 

I did rebuild GCC but it was before adding graphite as a flag, I'll get it done.

**Comment by fcsager** [comment]
Yes, you need to enable gcc's graphite USE flag, but the issue maybe that the glibc ebuild is not picking up CFLAGS from make.conf after sourcing make.conf.lto and/or make.conf.lto.defines, even with custom-cflags USE enabled.

sys-libs/glibc-2.32-r2 was failing to build for me with this error:

> In file included from <command-line>:
./../include/libc-symbols.h:75:3: error: #error "glibc cannot be compiled without optimization"
   75 | # error "glibc cannot be compiled without optimization"
      |   ^~~~~

Reviewing the build log showed my basic CFLAGS and CXXFLAGS from my make.conf were being used, but emerge --info glibc was showing the appropriate flags being set though:
```
=================================================================
                        Package Settings
=================================================================
sys-libs/glibc-2.32-r2::gentoo was built with the following:
USE="(crypt) custom-cflags multiarch (multilib) ssp (static-libs) -audit -caps (-cet) -compile-locales -doc -gd -headers-only -nscd -profile (-selinux) -static-pie -suid -systemtap -test (-vanilla)" ABI_X86="(64)"
CFLAGS="-march=icelake-client -O3 -fgraphite-identity -floop-nest-optimize -fdevirtualize-at-ltrans -fipa-pta -fuse-linker-plugin -pipe -mabm -mavx256-split-unaligned-load -mavx256-split-unaligned-store -mno-clwb -falign-functions=32 -Wl,-O1 -Wl,--as-needed"
CXXFLAGS="-march=icelake-client -O3 -fgraphite-identity -floop-nest-optimize -fdevirtualize-at-ltrans -fipa-pta -fuse-linker-plugin -pipe -mabm -mavx256-split-unaligned-load -mavx256-split-unaligned-store -mno-clwb -falign-functions=32 -Wl,-O1 -Wl,--as-needed"
FEATURES="split-elog buildpkg multilib-strict xattr ebuild-locks parallel-fetch ipc-sandbox binpkg-dostrip sfperms binpkg-docompress preserve-libs distlocks unmerge-orphans sandbox fixlafiles config-protect-if-modified qa-unresolved-soname-deps userfetch usersync userpriv assume-digests merge-sync unknown-features-warn pid-sandbox protect-owned unmerge-logs news network-sandbox ccache strict binpkg-logs usersandbox"
```

It was not until I copied CFLAGS and CXXFLAGS from emerge --info, pasted them into a file I created in /etc/portage/env, and referenced that file in package.env, I was able to get glibc to build again. My config is quoted below.

/etc/portage/env/glibc.conf (adjust as needed to match your system)
```
CFLAGS="-march=icelake-client -O3 -fgraphite-identity -floop-nest-optimize -fdevirtualize-at-ltrans -fipa-pta -fuse-linker-plugin -pipe -mabm -mavx256-split-unaligned-load -mavx256-split-unaligned-store -mno-clwb -falign-functions=32 -Wl,-O1 -Wl,--as-needed"
CXXFLAGS="-march=icelake-client -O3 -fgraphite-identity -floop-nest-optimize -fdevirtualize-at-ltrans -fipa-pta -fuse-linker-plugin -pipe -mabm -mavx256-split-unaligned-load -mavx256-split-unaligned-store -mno-clwb -falign-functions=32 -Wl,-O1 -Wl,--as-needed"
```
/etc/portage/package.env/glibc or append to /etc/portage/package.env
`sys-libs/glibc  glibc.conf`

**Comment by addeps3** [comment]
Me and the helpful friend xartin on discord have found my fstab file to contain incorrect syntax, 
Because Im a sloppy typer I had left out a space between the <dump> and  <fsck> columns on my home drive. 

I assumed this was a gentoo-LTO fault, but I could have just as well went to mainline gentoo community for support. 
Sorries! 
I'll close this and go bother them instead :)

**Comment by addeps3** [comment]
Not gonna re-open, just reporting that selecting default GCC version seems to have resolved my issue.

**Comment by telans** [comment]
I don't think you would have an issue if you try to emerge the latest glibc with GCC 10.2.0, it is generally discouraged to mix testing/stable packages

**Comment by addeps3** [comment]
I closed the issue before I saw the reply fcsager had posted. 

I'll try that one day. 
And apropos, glibc-2.32 didnt want to build, telans. 

I went ahead and rebuild my system with default repo's, but the -flto, -O3 and graphite flags remained in make.conf (forgot to change them) so I have a fully built system, just not with GCC 10. 
Oh, well. This system is just for testing, so its fine.

**Comment by fcsager** [comment]
> I closed the issue before I saw the reply fcsager had posted.

I posted that before finding out the effect of multiple -O flags in my cflags, and looking at ltoworkarounds.conf.

After running into some issues with other packages, I read up on the gcc -O flag and found that if multiple -O flags are specified, the last one is utilized, with the others ignored. So my issue would have been solved by simply editing /etc/make.conf and putting a -O2, for example, before the ${CFLAGS} in CFLAGS.

Secondly, ltoworkarounds.conf has glibc listed as "sys-libs/glibc *FLAGS-=-flto* #Alternative: *FLAGS+=-ffat-lto-objects *FLAGS-="${IPAPTA}" Fails in configure stage with LTO enabled, checking for builtin redirect. With -ffat-lto-objects, and no -fipa-pta, we get further, but get multiple definition errors  Forcing past that yields undefined symbol errors.". So glibc is a no LTO pacage. 

That said, glibc needs to have compiler optimization enabled by having a -O flag in your CFLAGS. I ended up forcing glibc to build though with LTO -O3 and graphite unbeknownst of the consequences. I'll continue to test it as I have built it, but I don't think it is recommended by the devs here.
